src/credentials/examples/DeviceAttestationVerifierExample.cpp

```diff
@@ -294,6 +328,67 @@ AttestationVerificationResult ExampleDACVerifier::ValidateCertificationDeclarati
     return AttestationVerificationResult::kSuccess;
 }
 
+AttestationVerificationResult
+ExampleDACVerifier::ValidateCertificateDeclarationPayload(const ByteSpan & certDeclBuffer, const ByteSpan & firmwareInfo,
+                                                          uint16_t clusterVendorId, uint16_t clusterProductId, uint16_t dacVendorId,
+                                                          uint16_t dacProductId, uint16_t paiProductId)
+{
+    CertificationElements decodedElements;
+    VerifyOrReturnError(DecodeCertificationElements(certDeclBuffer, decodedElements) == CHIP_NO_ERROR,
```

Comment:
This may overrun on non-compliant CD:

```
    certElements.ProductIdsCount = 0;
    while ((err = reader.Next(AnonymousTag)) == CHIP_NO_ERROR)
    {
        ReturnErrorOnFailure(reader.Get(certElements.ProductIds[certElements.ProductIdsCount++]));
    }
```

Neither the following two constants are used in DecodeCertificationElements:

```
constexpr uint32_t kCertificationElements_TLVEncodedMaxLength;
constexpr uint32_t kMaxCMSSignedCDMessage;
```

Therefore, lengths are trusted and it's possible to go past the bounds of the CD. 

These checks must be added now.

I would argue that an iterator for the CD's ProductId list would be easier. Therefore, I suggest replacing `DecodeCertificationElements` with a `CertificationElemeentsDecoder` class that has "immediate" extraction of the basic values and then a "containsPid" for the PID list. This would avoid needing to fully parse, maintain a max-sized stack view of the structure and would allow easier bound checking.